dango/client/src/signer.rs

```diff
@@ -1,22 +1,37 @@
 use {
     crate::SigningKey,
+    alloy::{
+        dyn_abi::{Eip712Domain, TypedData},
+        primitives::U160,
+    },
     bip32::{Language, Mnemonic},
+    dango_auth::EIP155_CHAIN_ID,
     dango_types::{
         account::spot,
         account_factory::Username,
-        auth::{Credential, Key, Metadata, Nonce, SignDoc, Signature, StandardCredential},
+        auth::{
+            Credential, Eip712Signature, Key, Metadata, Nonce, SignDoc, Signature,
+            StandardCredential,
+        },
         signer::SequencedSigner,
     },
     grug::{
-        Addr, Addressable, ByteArray, Defined, Hash256, HashExt, JsonSerExt, MaybeDefined, Message,
-        NonEmpty, QueryClient, QueryClientExt, SignData, Signer, StdResult, Tx, Undefined,
-        UnsignedTx,
+        Addr, Addressable, ByteArray, Defined, Hash256, HashExt, Inner, JsonDeExt, JsonSerExt,
+        MaybeDefined, Message, NonEmpty, QueryClient, QueryClientExt, SignData, Signer, StdError,
+        StdResult, Tx, Undefined, UnsignedTx, json,
     },
+    grug_crypto::keccak256,
     std::str::FromStr,
 };
 
 pub const DEFAULT_DERIVATION_PATH: &str = "m/44'/60'/0'/0/0";
 
+#[derive(Debug)]
+pub enum CredentialType {
+    Secp256k1,
+    Ethereum,
+}
```

Comment:
@Rhaki Instead of introducing this new enum type, I think it's better you create these new functions:

```rust
impl SingleSigner<Undefined<Nonce>> {
    pub fn new_secp256k1(
        username: &str,
        address: Addr,
        sk: SigningKey,
    ) -> anyhow::Result<Self> { /* ... */ }

    pub fn new_ethereum(
        username: &str,
        address: Addr,
        sk: SigningKey,
    ) -> anyhow::Result<Self> { /* ... */ }

    pub fn new_random_secp256k1(
        username: &str,
        address: Addr,
    ) -> anyhow::Result<Self> { /* ... */ }

    pub fn new_random_ethereum(
        username: &str,
        address: Addr,
    ) -> anyhow::Result<Self> { /* ... */ }
}
```

This way, you avoid introducing a new type.